bika/lims/utils/analysisrequest.py

```diff
@@ -549,3 +465,88 @@ def resolve_rejection_reasons(values):
         return [{"selected": selected, "other": other}]
 
     return []
+
+
+def do_rejection(sample, notify=None):
+    """Rejects the sample and if succeeds, generates the rejection pdf and
+    sends a notification email. If notify is None, the notification email will
+    only be sent if the setting in Setup is enabled
+    """
+    sample_id = api.get_id(sample)
+    if not sample.getRejectionReasons():
+        logger.warn("Cannot reject {} w/o rejection reasons".format(sample_id))
+        return
+
+    success, msg = doActionFor(sample, "reject")
+    if not success:
+        logger.warn("Cannot reject the sample {}".format(sample_id))
+        return
+
+    # Generate a pdf with the rejection reasons
+    pdf = get_rejection_pdf(sample)
+
+    # Attach the PDF to the sample
+    filename = "{}-rejected.pdf".format(sample_id)
+    sample.createAttachment(pdf, filename=filename)
+
+    # Do we need to send a notification email?
+    if notify is None:
+        setup = api.get_setup()
+        notify = setup.getNotifyOnSampleRejection()
+
+    if notify:
+        # Compose and send the email
+        mime_msg = get_rejection_mail(sample, pdf)
+        # Send the email
+        send_email(mime_msg)
+
+
+def get_rejection_pdf(sample):
+    """Generates a pdf with sample rejection reasons
+    """
+    # Avoid circular dependencies
+    from bika.lims.browser.analysisrequest.reject import \
+        AnalysisRequestRejectPdfView
+
+    # Render the html's rejection document
+    tpl = AnalysisRequestRejectPdfView(sample, api.get_request())
+    html = tpl.template()
+    html = safe_unicode(html).encode("utf-8")
+
+    # Generate the pdf
+    return createPdf(htmlreport=html)
+
+
+def get_rejection_mail(sample, rejection_pdf=None):
+    """Generates an email to sample contacts with rejection reasons
+    """
+    # Avoid circular dependencies
+    from bika.lims.browser.analysisrequest.reject import \
+        AnalysisRequestRejectEmailView
+
+    # Render the email body
+    tpl = AnalysisRequestRejectEmailView(sample, api.get_request())
+    email_body = tpl.template()
```

Comment:
Note there is another PR (to be reviewed/merged after this one) in which the generation of the email template does not longer rely on a BrowserView